README.md
# OpenVenue Backend Application
__OpenVenue__: A Map based application to help renting a house, built with Mapbox, Vue.js and loopback as backend server. http://openvenues.charlieli.cn/ 
[![Build Status](https://travis-ci.org/CCharlieLi/openVenue-backend.svg?branch=master)](https://travis-ci.org/CCharlieLi/openVenue-backend)
[![Coverage Status](https://coveralls.io/repos/github/CCharlieLi/openVenue-backend/badge.svg?branch=master)](https://coveralls.io/github/CCharlieLi/openVenue-backend?branch=master)

# Test

1. `npm i`
2. `./test_config/setupRedisForLocal.sh`
3. `npm test`
4. `./test_config/killRedisForLocal.sh`

# License
MIT

# Summary
```
 project  : openVenue-backend
 repo age : 4 months
 active   : 7 days
 commits  : 10
 files    : 29
 authors  : 
    10	CCharlieLi  100.0%
``` 


History.md

0.0.2 / 2016-08-16
==================

  * update package.json readme
  * set Travis CI and coverall
  * fix bug for updating venues, add tests, add test script

0.0.1 / 2016-08-16
==================

  * add test config
  * add update venue
  * add findVenue
  * use venue model
  * add user API
  * edit readme
  * deploy loopback
  * init


server/config.test.js
'use strict';

module.exports = {
  secretKey: '!Q@W#E$R'
};


server/test/venue.test.js
'use strict';

const chai = require('chai');
var chaiAsPromised = require('chai-as-promised');
chai.use(chaiAsPromised);
const should = chai.should();
const app = require('../server');
const request = require('supertest')(app);
const Venue = app.models.Venue;

const data = JSON.stringify({
  data: {
    geoHash: '123456a',
    username: 'Charlie',
    password: '123456',
    wechat: '449217425',
    venueName: 'qwwe',
    other: '123',
    coordinate: {
      lng: 120,
      lat: 31
    }
  }
});

const fdata = JSON.stringify({
  data: {
    geoHash: '123456a',
    username: 'Charlie',
    password: '1234567',
    wechat: '449217425',
    venueName: 'qwwe',
    other: '123',
    coordinate: {
      lng: 120,
      lat: 31
    }
  }
});

const brokenData = JSON.stringify({
  data: {
    username: 'Charlie',
    password: '123456',
    venueName: 'qwwe',
    other: '123'
  }
});

describe('Venue API', () => {
  describe('Add Venue', () => {
    it('should return 200 when sign up data', function (done) {
      request.post('/venues/addvenue')
        .type('application/json')
        .send(data)
        .expect(200)
        .end((err, res) => {
          let data = res.body.data;
          data.Password.should.be.equal('');
          data.should.have.ownProperty('id');
          data.should.have.ownProperty('Username');
          data.should.have.ownProperty('VenueName');
          data.should.have.ownProperty('id_token');
          done(err);
        });
    });

    it('should return 200 when update data', function (done) {
      request.post('/venues/addvenue')
        .type('application/json')
        .send(data)
        .expect(200)
        .end((err, res) => {
          let data = res.body.data;
          data.Password.should.be.equal('');
          data.should.have.ownProperty('id');
          data.should.have.ownProperty('Username');
          data.should.have.ownProperty('VenueName');
          data.should.have.ownProperty('id_token');
          done(err);
        });
    });

    it('should return 403 update with wrong password', function (done) {
      request.post('/venues/addvenue')
        .type('application/json')
        .send(fdata)
        .expect(403)
        .end(done);
    });

    it('should return 400 when sign up with missing data', function (done) {
      request.post('/venues/addvenue')
        .type('application/json')
        .send(brokenData)
        .expect(400)
        .end(done);
    });

    it('should return 400 when sign up with wrong format', function (done) {
      request.post('/venues/addvenue')
        .type('application/json')
        .send(JSON.stringify({
          username: 'Charlie',
          password: '123456',
          wechat: '449217425',
          venueName: 'qwwe',
          other: '123'
        }))
        .expect(400)
        .end(done);
    });
  });

  describe('Find Venue', () => {
    // beforeEach('Flush redis', (done) => {
    //   Venue.getConnector().connect().call('flushall').then(function () {
    //     done();
    //   }, done);
    // });

    // after('Flush redis', (done) => {
    //   Venue.getConnector().connect().call('flushall').then(function () {
    //     done();
    //   }, done);
    // });

    it('should return 200 when find a saved venue', function (done) {
      request.post('/venues/findvenue')
        .type('application/json')
        .send(JSON.stringify({
          data: {
            geoHash: '123456a'
          }
        }))
        .expect(200)
        .end((err, res) => {
          let data = res.body.data;
          data.username.should.be.equal('Charlie');
          done(err);
        });
    });

    it('should return 200 when find all saved venue', function (done) {
      request.get('/venues/findallvenues')
        .expect(200)
        .end((err, res) => {
          let data = res.body.data;
          data.should.be.Array;
          done(err);
        });
    });
  });

  describe('DELETE Venue', () => {
    // beforeEach('Flush redis', (done) => {
    //   Venue.getConnector().connect().call('flushall').then(function () {
    //     done();
    //   }, done);
    // });

    // after('Flush redis', (done) => {
    //   Venue.getConnector().connect().call('flushall').then(function () {
    //     done();
    //   }, done);
    // });

    it('should return 200 when delete venue', function (done) {
      request.post('/venues/deletevenue')
        .type('application/json')

        // .set('Authorization', 'Token token=blahblahblah')
        .send(JSON.stringify({
          data: {
            geoHash: '123456a',
            venueName: 'qwwe',
            password: '123456'
          }
        }))
        .expect(200)
        .end(done);
    });
  });
});


server/datasources.test.js
'use strice;';

module.exports = {
  redisDs: {
    connector: 'ioredis',
    name: 'redisDs',
    host: 'localhost',
    port: 6379
  }
};


common/models/person.js
'use strict';
const Promise = require('bluebird');
const httpError = require('http-errors');
const utils = require('../utils');

module.exports = (Person) => {

  Person.addUser = (data) => {
    return new Promise((resolve, reject) => {
      if (!data.username || !data.password || !data.wechat || !data.venueName || !data.other) {
        reject(httpError(400, 'sign with missing data.'));
      }
      Person.create({
        id: data.venueName,
        Username: data.username,
        Password: data.password,  // TODO: Password encryption
        Wechat: data.wechat,
        VenueName: data.venueName,
        Other: data.other
      }).then((res) => {
        resolve({
          id_token: utils.createToken(data.password)
        });
      }, (err) => {
        reject(err);
      });
    });
  };

  Person.deleteUser = (auth, data) => {
    //TODO: auth validation
    return new Promise((resolve, reject) => {
      Person.destroyById(data.username).then(function (res) {
        if (res.count == 1) {
          resolve(httpError(200, 'Account deleted.'));
        }
        reject(httpError(404, 'Account not found.'));
      });
    });
  };

  Person.remoteMethod('addUser', {
    description: 'sign up',
    http: { path: '/addvenue', verb: 'post' },
    returns: { type: 'object', arg: 'data' },
    accepts: [

      // { arg: 'auth', type: 'string', required: true, http: (ctx) => ctx.req.headers['Authorization'] },
      { arg: 'data', type: 'object', required: true }
    ]
  });

  Person.remoteMethod('deleteUser', {
    description: 'delete account',
    http: { path: '/deletevenue', verb: 'post' },
    returns: { type: 'object', arg: 'data' },
    accepts: [
      { arg: 'auth', type: 'string', required: true, http: (ctx) => ctx.req.headers.authorization },
      { arg: 'data', type: 'object', required: true }
    ]
  });
};


common/models/venue.js
'use strict';
const Promise = require('bluebird');
const httpError = require('http-errors');
const utils = require('../utils');

module.exports = (Venue) => {

  Venue.addVenue = (data) => {
    if (!data.username || !data.password || !data.wechat || !data.venueName ||
     !data.other || !data.coordinate || !data.geoHash) {
      return Promise.reject(httpError(400, 'sign with missing data.'));
    }
    return Venue.findById(data.geoHash).then(function (res) {

      if (res === null || res.Password === data.password.toString()) {
        return Venue.updateOrCreate({
          id: data.geoHash,
          Username: data.username,
          Password: data.password,  // TODO: Password encryption
          Wechat: data.wechat,
          VenueName: data.venueName,
          Other: data.other,
          Coordinate: data.coordinate
        }).then((res) => {
          res.Password = '';
          return Object.assign(res, { id_token: utils.createToken(data.password) });
        }, (err) => {
          return err;
        });
      }
      return Promise.reject(httpError(403, 'wrong password.'));
    });
  };

  Venue.findVenue = (data) => {
    if (!data.geoHash) {
      reject(httpError(400, 'sign with missing data.'));
    }
    return new Promise((resolve, reject) => {
      Venue.findById(data.geoHash).then(function (res) {
        if (res) {
          resolve({
            username: res.Username,
            wechat: res.Wechat,
            geoHash: res.id,
            venueName: res.VenueName,
            other: res.Other,
            coordinate: JSON.parse(res.Coordinate)
          });
        }
        reject(httpError(404));
      });
    });
  };

  Venue.findAllVenues = () => {
    return new Promise((resolve, reject) => {
      Venue.find().then(function (res) {
        let ret = res.map((re) => {
          return {
            username: re.Username,
            wechat: re.Wechat,
            geoHash: re.id,
            venueName: re.VenueName,
            other: re.Other,
            coordinate: JSON.parse(re.Coordinate)
          };
        });
        resolve(ret);
      });
    });
  };

  Venue.deleteVenue = (data) => {
    if (!data.password || !data.geoHash) {
      reject(httpError(400, 'sign with missing data.'));
    }
    return new Promise((resolve, reject) => {
      Venue.findById(data.geoHash).then(function (res) {
        if (res.Password === data.password.toString()) {
          Venue.destroyById(data.geoHash).then(function (re) {
            if (re.count === 1) {
              resolve(httpError(200, 'Venue deleted.'));
            }
          });
        } else {
          reject(httpError(404, 'Account not found.'));
        }
      });
    });
  };

  Venue.remoteMethod('addVenue', {
    description: 'add venue',
    http: { path: '/addvenue', verb: 'post' },
    returns: { type: 'object', arg: 'data' },
    accepts: [

      // { arg: 'auth', type: 'string', required: true, http: (ctx) => ctx.req.headers['Authorization'] },
      { arg: 'data', type: 'object', required: true }
    ]
  });

  Venue.remoteMethod('findVenue', {
    description: 'find one venue',
    http: { path: '/findvenue', verb: 'post' },
    returns: { type: 'object', arg: 'data' },
    accepts: [

      // { arg: 'auth', type: 'string', required: true, http: (ctx) => ctx.req.headers.authorization },
      { arg: 'data', type: 'object', required: true }
    ]
  });

  Venue.remoteMethod('findAllVenues', {
    description: 'find all venues',
    http: { path: '/findallvenues', verb: 'get' },
    returns: { type: 'object', arg: 'data' }
  });

  Venue.remoteMethod('deleteVenue', {
    description: 'delete venue',
    http: { path: '/deletevenue', verb: 'post' },
    returns: { type: 'object', arg: 'data' },
    accepts: [

      // { arg: 'auth', type: 'string', required: true, http: (ctx) => ctx.req.headers.authorization },
      { arg: 'data', type: 'object', required: true }
    ]
  });
};
